### ISSUE TITLE
Discussion: Rip out Request Parser and marshalling system and replace them with Marshmallow

### ISSUE THREAD

**Issue by svenstaro** [flask-restful/flask-restful#335]
Hi, since I discovered flask-restful I've been wary of its marshalling support and the under-featured Request Parser. In short, I want to replace those with [Marshmallow](http://marshmallow.readthedocs.org/en/latest/) which specializes on doing that exact thing and it does it very well. This might not be for 1.0.0, however. I don't want to break everyone's applications.
# Why?

I want to do this for two reason mainly. Firstly, most issues in this project are either with the Request Parser (#275 #258 #279 #326 #322 #321 #317 #305 #301) or the marshalling support (#291 #309 #332 #324 #318 #300). Secondly, since we already have tools in the Python world to do all that for us, I'm not sure we should reinvent the wheel badly and put it into this library. It would serve everybody to use proper solutions for this and to outsource the effort required for maintaining those.

Now, obviously we might go one step further and just propose to have entirely generic solutions in place of this with documentation on how to integrate them. I'm skeptical of that as it might lead to overengineering. I think by using Marshmallow (which actually already handles most of the current issues that we have with marshalling and validation) we can get rid of a lot of code as well.

I know this is a radical change so I want to discuss it with you guys first.

**Comment by taion** [comment]
This sounds great for marshalling.

How does parsing e.g. `request.args` for a `GET` request work in this case, though? Would the idea be to use webargs instead of reqparse? It seems something like what is present right now is more expressive for this use case than what marshmallow offers.

**Comment by taion** [comment]
This might be a dumb novice question, but what is actually left of Flask-RESTful beyond what is offered by `MethodView` after you strip out the marshalling and request parsing code?

Judging by the docs and the code for `Resource`, the only things available are `method_decorators` on `Resource`, which mirrors functionality actually available on `MethodView`, and the code for handling media types. While the latter in undoubtedly useful, it's still a somewhat limited thing. Net of the marshalling and request parsing, does the average use case get much out of using Flask-RESTful over just extending Flask `MethodView`s?

**Comment by justanr** [comment]
@taion I think both of those are the real value of Restful, not the request parameter parsing and marshalling. 

Flask-Classy is also a thin wrapper around MethodView (or maybe it's just View, I don't recall off hand), but it has seen success as well. Though, more of a kindof Blueprint replacement than an API tool.

Tearing these two portions out restricts the scope of the project, but that's also not a bad thing. Especially because Marshmallow is much more robust (see self nesting serializers between the two tools). 

I'm +1 on this.

**Comment by keisetsu** [comment]
Are there plans to note this possible deprecation in the documentation? I realize this is pre-1.0, but I've seen mention of dumping reqparse in a few issues, and it'd be good to note it so that people (like me :) ) don't spend time on it.

**Comment by svenstaro** [comment]
@keisetsu I added a note about it to the docs.

**Comment by EFF** [comment]
Where are we at this point ? Is Swapping RequestParser for Marshmallow on the roadmap?

I have some changes in the that might be useful for the current RequestParser but if the plan is to swap it (which I think is a good idea), I'm not  going open any PR.

**Comment by taion** [comment]
If I'm not mistaken, it'd actually be:

RequestParser -> webargs
marshal -> marshmallow

No? You could use marshmallow to parse arguments (and should for deserializing incoming models), but webargs is explicitly the tool (and from the same author, at that) for doing so.

**Comment by svenstaro** [comment]
@EFF Yes, that is the idea. We will introduce new docs that show off how to use flask-restful with marshmallow.
@taion Yes, indeed.

**Comment by EFF** [comment]
Seems legit to me! I could help on tackling this issue if not started !

**Comment by svenstaro** [comment]
@EFF That would be great. So, the idea is: Rip out EVERYTHING to do with the parser in flask-restful, docs, code, examples. Replace with examples and docs on how to do that stuff (but better) with marshmallow and webargs in examples from simple to intermediate.

**Comment by taion** [comment]
@svenstaro This is echoing an earlier point a bit, but as a user, I'd find it more helpful for the docs to focus on what Flask-RESTful actually _does_ outside of the argument parser and serialization bits. If I'm writing an API that speaks JSON, it's not clear to me why I should use Flask-RESTful at all, if I'm using webargs and marshmallow.

**Comment by EFF** [comment]
I agree with @taion, it would be hard to keep examples and docs up-to-date with webargs and marshmallow since they are external libs. I would definitely refer to these libs in the docs though. Then, correct me if I'm wrong, flask-restful becomes a lib that helps you to organise your routes in resources.

**Comment by svenstaro** [comment]
Fair enough, guys. In that case, though, this extension is going to become very slim. Maybe that's a good thing so it could focus on that. I agree referring to extensions might be a better thing, however, I think that keeping the examples up to date if likely not an issue since this is Github after all and the extensions don't move too quickly. It might just be helpful to have something along to lines of "How to actually build something real with flask-restful".

It might also be worth considering if flask-restful is actually required at all in the ecosystem. I mean, all things it does, other extensions nowadays do better (but weren't an option back when it was created). For instance, flask-classy + marshmallow + webargs is a better stack than flask-restful.

What do you guys think?

**Comment by taion** [comment]
Thoughts:

* If you're using doing JSON, `MethodView` + those packages gets you something very similar to what Flask-RESTful currently offers.
* Definitely the biggest missing thing without e.g. Flask-Classy is the ability to define both e.g. `/api/users` and `/api/users/<id>` as the same endpoint. However, Flask-Classy does more than that - I think there's room for something very lightweight that just gives a basic abstraction around REST resources with operations on both the collection and the item level. I'm happy enough just to set up entirely separate resources for subresources, as opposed to doing things the Flask-Classy way.
* I think the content type stuff in Flask-RESTful is going to be extremely useful to people looking to scale up their APIs to, uhh, handle multiple response content types. If I'm not mistaken, Flask-Classy doesn't take care of this sort of thing at all.
* For CRUD-style operations, something like the mixin-based scheme for building views from DRF would be extremely useful, but nothing in the Flask ecosystem except for Flask-Admin really has this kind of thing, and it's really annoying. You have either extremely lightweight packages like Flask-RESTful and Flask-Classy that give you almost no help at all, or extremely heavyweight packages like Flask-Restless and Eve that give you extremely limited code-based flexibility for defining views. I really wish there were something like DRF that would let me just add mixins to a `MethodView`-alike to add basic support for relevant methods.

**Comment by ColeKettler** [comment]
I think it's worth noting that flask-classy looks pretty dead. No commits / response for almost a year and a half.

**Comment by sloria** [comment]
Hello! Maintainer of marshmallow and webargs here. First of all, I want to thank you for your work on this library; my libs were heavily inspired by Flask-RESTful.

Just wanted to mention that both libs recently had stable releases, making this good time to document the integration. As a starting point, there is an example of using webargs with Flask-RESTful here: https://github.com/sloria/webargs/blob/dev/examples/flaskrestful_example.py .

**Comment by jo-tham** [comment]
nice start @sloria & thanks to the flask-restful contributors for work so far; the project is pretty helpful

@sloria, can you comment on how to include a field as a url to another endpoint as with #318 ? If so, I would start documenting the integration as I'm at that stage of constructing a service with flask restful now.

and what's the implication for flask-restful, then? It would just be for extending views and mapping resources routes? sounds like a reasonable isolation of responsibilities.

**Comment by jo-tham** [comment]
after 2nd inspection and some other reading, im with @taion 's previous post and i found @sloria  explanation of marshalling resource/endpoint urls at https://github.com/marshmallow-code/flask-marshmallow...

**Comment by sloria** [comment]
I encourage people to take a look at [flask-apispec](https://github.com/jmcarp/flask-apispec), which has a similar API to Flask-RESTful's `marshal_with`. It supports input/output with webargs and marshmallow and adds automatic swagger documentation.

I am considering merging flask-apispec with flask-marshmallow if there is interest. I welcome your feedback on https://github.com/marshmallow-code/flask-marshmallow/issues/26.

**Comment by imaia** [comment]
+1 for marshmallow
I agree that having flask-restful focused on restful API problems, would make it even more useful and would avoid functionality "clash" with other popular APIs.

**Comment by jeffwidman** [comment]
:+1: for using `marshmallow`/`flask-marshmallow` rather than custom-baked marshalling/serializing. 

I've been really happy with `flask-marshmallow` on some of my projects.

**Comment by henadzit** [comment]
I would suggest to refer marshmallow and webargs at [this](http://flask-restful-cn.readthedocs.org/en/0.3.4/reqparse.html) and [this](http://flask-restful-cn.readthedocs.org/en/0.3.4/fields.html) pages as more flexible alternatives for now. 

If you're a newcomer and you start building your project with flask-restful, you soon realize that there're lots of limitations of regparse and marshal. But at the moment when you realize that, some code and some hacks have been written and they need to be redone. Talking from my personal experience...

I like marshmallow so far. Actually, it doesn't require any supporting code to use it along with flask-restful except returning error code explicitly. Personally, I don't like flask-marshmallow because it makes you to import the ma object (ma = Marshmallow(app)) in all resource modules.

**Comment by joshfriend** [comment]
@henadzit what are those links? Isn't that just the Flask-RESTful docs mirrored to a different RTD page?

**Comment by henadzit** [comment]
@joshfriend oh, yes, they're. I meant to put links to the original documentation.

**Comment by lafrech** [comment]
@ColeKettler:

> I think it's worth noting that flask-classy looks pretty dead. No commits / response for almost a year and a half.

Indeed: https://github.com/apiguy/flask-classy/issues/80.

There's an active fork called [flask-classful](https://github.com/teracyhq/flask-classful).

- http://flask-classful.teracy.org/
- https://pypi.python.org/pypi/Flask-Classful

**Comment by bradwood** [comment]
Hey guys, did this conversation get anywhere? 

Has reqparse been finally ripped out? 

For someone about to to commit to flask-restful, what is the current best practice recommendation for handling complex JSON payloads?

**Comment by svenstaro** [comment]
The problem with actually doing this is that it would make this project effectively a wrapper around flask's routing system. At this point it might just be best to declare the whole project deprecated but maintained and refer people to projects and guides that easily replace all of this with more modern tools. Thoughts?

**Comment by taion** [comment]
I would say that a lot of what Flask-RESTful does is better handled by composing more atomic libraries – essentially just using Marshmallow plus Flask `MethodView`s gets you a lot of the way. Content negotiation is something unique to Flask-RESTful, though – but many APIs don't need that.

**Comment by 1Mark** [comment]
any progress on this issue?
